### PR TITLE
VarCouldBeVal: don't report assigned property that is defined in anonymous object

### LINKED ISSUE #3805: Wrong VarCouldBeVal report
## Expected Behavior

I shouldn't get a report about `VarCouldBeVal` for a property in an object in a function.

## Observed Behavior

I get a `VarCouldBeVal` issue for a property, that cannot be changed to `val`

## Steps to Reproduce

The following code reproduces the issue
```kotlin
fun test(){
    val wrapper = object {
        var test: Boolean = true
    }
    wrapper.test = false
}
```

## Your Environment

* Version of detekt used: 1.17.0
* Version of Gradle used (if applicable): 7.0.2
* Gradle scan link (add `--scan` option when running the gradle task):
* Operating System and version:
* Link to your project (if it's a public repository): https://github.com/EdwarDDay/serialization.kprefs/pull/98

### PR DESCRIPTION
**Pull request by t-kameyama**
Fixes #3805

### CODE REVIEW

**Review by BraisGabin — APPROVED** [review]

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/VarCouldBeVal.kt [hunk 1]
```diff
@@ -41,6 +43,7 @@ private val unaryAssignmentOperators = setOf(KtTokens.MINUSMINUS, KtTokens.PLUSP
  * </compliant>
  */
 @ActiveByDefault(since = "1.16.0")
+@RequiresTypeResolution
```
Comment (chao2zhang) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/VarCouldBeVal.kt:46]: It seems quite unfortunate that we are now converting plain rules to type resolving rules like such.

I think this PR should still be good to go, but I would want to highlight this for us to favor detekt compiler plugin.

**Review by chao2zhang — APPROVED** [review]

**Review by schalkms — APPROVED** [review]

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/VarCouldBeVal.kt (hunk 1 above)
Comment (cortinico) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/VarCouldBeVal.kt:46]: I believe we could mention this in the release notes but we're good to go imho.

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/VarCouldBeValSpec.kt [hunk 2]
```diff
@@ -144,11 +149,34 @@ class VarCouldBeValSpec : Spek({
...
-                // we accept wrong entity reporting here due to no type resolution
-                // false reporting with shadowed vars vs false positives
-                assertThat(entity.ktElement?.text).isEqualTo("private var myVar: String? = null")
+            with(subject.compileAndLintWithContext(env, code)[0]) {
+                assertThat(entity.ktElement?.text).isEqualTo("var myVar = value")
             }
         }
     }
+
+    describe("properties defined in anonymous object - #3805") {
```
Comment (cortinico) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/VarCouldBeValSpec.kt:158]: Would you be able to add a test case also for:
https://github.com/detekt/detekt/issues/3805#issuecomment-862068331

**Review by cortinico — COMMENTED** [review]
Great work, thanks for fixing this 🙏

**Review by cortinico — APPROVED** [review]

**Pull request merged**